Kernel/Devices/PATAChannel.cpp

```diff
@@ -98,34 +100,31 @@ PATAChannel::PATAChannel(ChannelType type)
     m_dma_enabled.resource() = true;
     ProcFS::add_sys_bool("ide_dma", m_dma_enabled);
 
-    initialize();
+    initialize(force_pio);
     detect_disks();
 }
 
 PATAChannel::~PATAChannel()
 {
 }
 
-void PATAChannel::initialize()
+void PATAChannel::initialize(bool force_pio)
 {
-    static const PCI::ID piix3_ide_id = { 0x8086, 0x7010 };
-    static const PCI::ID piix4_ide_id = { 0x8086, 0x7111 };
     PCI::enumerate_all([this](const PCI::Address& address, PCI::ID id) {
-        if (id == piix3_ide_id || id == piix4_ide_id) {
+        if (PCI::get_class(address) == PCI_Mass_Storage_Class && PCI::get_subclass(address) == PCI_IDE_Controller_Subclass) {
             m_pci_address = address;
-            kprintf("PATAChannel: PIIX%u PATA Controller found!\n", id == piix3_ide_id ? 3 : 4);
+            kprintf("PATAChannel: PATA Controller found! id=%w:%w\n", id.vendor_id, id.device_id);
         }
     });
-
     // Let's try to set up DMA transfers.
-    if (!m_pci_address.is_null()) {
+    if (!force_pio) {
```

Comment:
This change seems wrong. `m_pci_address` can still be null if we don't find an IDE controller on the PCI bus.

Reply:
I fixed that before you mentioned it, however that doesn't allow me to boot with a partitioned image (which is a bug I'll look into later)...